crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -214,4 +208,81 @@ impl QueryLimitsChecker {
 
         Ok(())
     }
+
+    fn analyze_selection_set(
+        &self,
+        cfg: &ServiceConfig,
+        fragment_defs: &HashMap<Name, Positioned<FragmentDefinition>>,
+        sel_set: &Positioned<SelectionSet>,
+    ) -> ServerResult<ComponentCost> {
+        // Use BFS to analyze the query and
+        // count the number of nodes and the depth of the query
+
+        // Queue to store the nodes at each level
+        let mut que = VecDeque::new();
+        // Number of nodes in the query
+        let mut num_nodes: u32 = 0;
+        // Depth of the query
+        let mut depth: u32 = 0;
+        // Number of nodes at each level
+        let mut level_len;
+
+        for top_level_sel in sel_set.node.items.iter() {
+            que.push_back(top_level_sel);
+            num_nodes += 1;
+            self.check_limits(cfg, num_nodes, depth, Some(top_level_sel.pos))?;
+        }
+
+        // Track the number of nodes at first level if any
+        level_len = que.len();
+
+        while !que.is_empty() {
+            // Signifies the start of a new level
+            depth += 1;
+            self.check_limits(cfg, num_nodes, depth, None)?;
```

Comment:
`check_limits` does not seem to need access to `self` -- consider making it a free function?